### PR TITLE
Add algebraic simplifier

### PR DESCRIPTION
**Pull request by parttimenerd**

### CODE REVIEW

File: src/main/java/minijava/ir/optimize/AlgebraicSimplifier.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  @Override
+  public boolean optimize(Graph graph) {
+    this.graph = graph;
+    hasChanged = false;
+    nodesToExchange = new HashMap<>();
+    BackEdges.enable(graph);
+    boolean hasChanged = fixedPointIteration();
+    replaceNodes();
+    BackEdges.disable(graph);
+    return true;
```
Comment (sherter) [inline src/main/java/minijava/ir/optimize/AlgebraicSimplifier.java]: Why are you always returning true?

File: src/main/java/minijava/ir/optimize/BaseOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import firm.BackEdges;
+import firm.Graph;
+import firm.nodes.Node;
+
+public abstract class BaseOptimizer extends DefaultNodeVisitor implements Optimizer {
+
+  protected Graph graph;
+  protected boolean hasChanged;
+
+  protected boolean fixedPointIteration() {
```
Comment (sherter) [inline src/main/java/minijava/ir/optimize/BaseOptimizer.java:19]: Please add some documentation for the boolean return value

File: src/main/java/minijava/ir/optimize/Optimizer.java [hunk 3]
```diff
@@ -3,5 +3,5 @@
 import firm.Graph;
 
 public interface Optimizer {
-  void optimize(Graph graph);
+  boolean optimize(Graph graph);
```
Comment (sherter) [inline src/main/java/minijava/ir/optimize/Optimizer.java:13]: Please add documentation for the boolean return value

File: src/main/java/minijava/ir/optimize/UnreachableCodeRemover.java [hunk 4]
```diff
@@ -6,11 +6,12 @@
...
 
   @Override
-  public void optimize(Graph graph) {
+  public boolean optimize(Graph graph) {
     // first remove Bad nodes we inserted manually (the remove_unreachable_code function doesn't pick them up)
     binding_irgopt.remove_bads(graph.ptr);
     binding_irgopt.remove_unreachable_code(
         graph.ptr); // find and replace unreachable code with Bad nodes
     binding_irgopt.remove_bads(graph.ptr);
+    return true;
```
Comment (sherter) [inline src/main/java/minijava/ir/optimize/UnreachableCodeRemover.java]: Isn't it possible to detect changes? Maybe `graph.getLastIdx()` or something?

**Review by sherter — CHANGES_REQUESTED** [review]

File: src/main/java/minijava/ir/optimize/AlgebraicSimplifier.java (hunk 1 above)
Comment (parttimenerd) [reply src/main/java/minijava/ir/optimize/AlgebraicSimplifier.java]: A mistake.

**Review by sgraf812 — APPROVED** [review]

File: src/main/java/minijava/ir/optimize/AlgebraicSimplifier.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+  @Override
+  public void visit(Mul node) {
+    Optional<Tuple2<Const, Node>> subNodesOpt = split(node);
+    if (!subNodesOpt.isPresent()) {
+      return;
+    }
+    Tuple2<Const, Node> subNodes = subNodesOpt.get();
+    if (subNodes.v1.getTarval().isNull()) { // x * 0  == 0
+      exchange(node, subNodes.v1);
```
Comment (sherter) [inline src/main/java/minijava/ir/optimize/AlgebraicSimplifier.java]: Calling this helper method `exchange` is misleading. My first thought was that you are mixing up analysis and optimization phase...

**Review by sherter — APPROVED** [review]

**Pull request merged**